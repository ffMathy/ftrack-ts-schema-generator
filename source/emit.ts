import type { QuerySchemasResponse, Session } from "@ftrack/api";
import * as fs from "fs";
import * as path from "path";
import prettier from "prettier";
import { convertSchemaToInterface } from "./convertSchemaToInterface.js";
import type CustomAttributeConfigurations from "./__snapshots__/responses/query_custom_attribute_configurations.json";

const legacySchemas = ["Conversation", "Message", "Participant"];
export async function emitToFile(
  session: Session,
  outputPath = "__generated__",
  outputFilename = "schema.ts"
) {
  // Get the schemas from the server and sort by id in alphabetical order
  const schemas = await getSchemas(session);
  const customAttributes = await getCustomAttributes(session);

  const {
    prettifiedContent,
    errors,
  }: { prettifiedContent: string; errors: unknown[] } = await emitToString(
    session.serverVersion,
    session.serverUrl,
    schemas,
    customAttributes
  );
  fs.mkdirSync(path.resolve(outputPath), { recursive: true });
  fs.writeFileSync(path.join(outputPath, outputFilename), prettifiedContent);

  return { errors, schemas };
}

type CustomAttributeConfiguration =
  (typeof CustomAttributeConfigurations)[number];

export async function emitToString(
  serverVersion: string | undefined,
  serverUrl: string | undefined,
  schemas: QuerySchemasResponse,
  customAttributes: CustomAttributeConfiguration[]
) {
  const preamble = `// :copyright: Copyright (c) ${new Date().getFullYear()} ftrack \n\n// Generated on ${new Date().toISOString()} using schema \n// from an instance running version ${serverVersion} using server on ${serverUrl} \n// Not intended to modify manually\n\n`;

  const errors: unknown[] = [];

  if (schemas.length < 1) {
    errors.push("No schemas found!");
    return {
      prettifiedContent: "",
      errors,
    };
  }

  // For each schema in schemas, convert it to a TypeScript interface and add to a string
  let interfaces = "";
  for (const schema of schemas) {
    if (legacySchemas.includes(schema.id)) {
      continue;
    }

    const { TSInterface, conversionErrors } = await convertSchemaToInterface(
      schema,
      schemas
    );
    errors.push(...conversionErrors);
    interfaces += TSInterface;
  }

  const customAttributesString = `
    export function getAttributeConfigurations() {
      return [
        ${customAttributes.map(
          (x) => `{
            name: "${x.key}",
            label: "${x.label}",
            entityType: "${x.entity_type}",
            objectType: ${x.object_type ? `"${x.object_type.name}"` : "undefined"},
            isHierarchical: ${x.is_hierarchical}
          }`
        )}
      ] as const;;
    }
    
    export type CustomAttributeConfiguration = ReturnType<typeof getAttributeConfigurations>[number];
    export type CustomAttributeConfigurationName = CustomAttributeConfiguration["name"];
    export type CustomAttributeConfigurationLabel = CustomAttributeConfiguration["label"];
  `;

  // Add a map of entity types and type for EntityType and a type for EntityData
  const schemaNames = schemas
    .map((s) => s.id)
    .map((name) => `${name}: ${name};`);
  const entityTypeMap = `export interface EntityTypeMap {${schemaNames.join(
    "\n"
  )}}`;
  const entityType = `export type EntityType = keyof EntityTypeMap;`;
  const entityData = `export type EntityData<TEntityType extends EntityType = EntityType> =
    EntityTypeMap[TEntityType];`;
  // Add a map of TypedContext subtypes and type for TypedContextSubtype
  const { TypedContextSubtypeMap, TypedContextSubtype } =
    getTypedContextTypes(schemas);
  // Write the file, adding the preamble first and the errors to the end of the file
  // Hack to get around the fact that the API doesn't return the correct type for BasicLink
  // Task: 95e02be2-c7ec-11ed-ae64-46416ff77027
  interfaces = AddBasicLinkType(interfaces);
  // End of hack
  // Prettify
  const allContent =
    preamble +
    interfaces +
    entityTypeMap +
    entityType +
    entityData +
    TypedContextSubtypeMap +
    TypedContextSubtype +
    customAttributesString +
    "\n \n// Errors: \n" +
    errors.map((error) => `// ${error}`).join("\n");
  const prettifiedContent = prettier.format(allContent, {
    parser: "typescript",
  });
  return { prettifiedContent, errors };
}

async function getCustomAttributes(session: Session) {
  const customAttributes = await session.query<CustomAttributeConfiguration>(
    "select label, values, key, project_id, entity_type, is_hierarchical, object_type.name from CustomAttributeConfiguration order by sort"
  );
  return customAttributes.data;
}

async function getSchemas(session: Session) {
  const schemas = await session.call<QuerySchemasResponse>([
    {
      action: "query_schemas",
    },
  ]);
  schemas[0].sort((a, b) => (a.id > b.id ? 1 : -1));
  return schemas[0];
}

function AddBasicLinkType(interfaces: string) {
  const regex = /link\?:\s*string/gi;
  const modifiedInterfaces = interfaces.replace(regex, "link?: BasicLink[]");
  return `${modifiedInterfaces} export interface BasicLink {
    id: string;
    type: string;
    name: string;
  };`;
}
function getTypedContextTypes(schemas: QuerySchemasResponse) {
  const typedContextNames = schemas
    .filter(
      (schema) =>
        (typeof schema?.alias_for === "object" &&
          schema.alias_for.id === "Task") ||
        schema.id === "TypedContext"
    )
    .map((s) => s.id)
    .map((name) => `${name}: ${name};`);
  const TypedContextSubtypeMap = `export interface TypedContextSubtypeMap {${typedContextNames.join(
    "\n"
  )}}`;
  const TypedContextSubtype = `export type TypedContextSubtype = keyof TypedContextSubtypeMap;`;
  return { TypedContextSubtypeMap, TypedContextSubtype };
}
