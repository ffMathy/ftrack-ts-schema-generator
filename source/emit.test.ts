import { expect, test, vi } from "vitest";
import { emitToString } from "./emit";
import { QuerySchemasResponse, Schema } from "@ftrack/api";

test("emitting with no schemas returns error", async () => {
  //act
  const emitResult = await emitToString(
    "4.13.8",
    "https://ftrack.example.com",
    []
  );

  //assert
  expect(emitResult?.errors[0]).toBe("No schemas found!");
});

test("schema subtype of TypedContext", async () => {
  //arrange
  vi.setSystemTime(new Date(2023, 1, 1));

  const schemas: Array<Partial<Schema>> = [
    getTypedContextSchema(),
    {
      id: "SomeInterfaceName",
      alias_for: {
        id: "Task",
      },
    },
  ];

  //act
  const emitResult = await emitToString(
    "4.13.8",
    "https://ftrack.example.com",
    schemas as QuerySchemasResponse
  );

  //assert
  expect(emitResult.errors).toHaveLength(0);
  expect(emitResult.prettifiedContent).toMatchSnapshot();
});

test("schema has base schema", async () => {
  //arrange
  vi.setSystemTime(new Date(2023, 1, 1));

  const schemas: Array<Partial<Schema>> = [
    getTypedContextSchema(),
    {
      id: "SomeInterfaceName",
      $mixin: { $ref: "SomeBaseSchema" },
    },
    {
      id: "SomeBaseSchema",
      type: "variable",
    },
  ];

  //act
  const emitResult = await emitToString(
    "4.13.8",
    "https://ftrack.example.com",
    schemas as QuerySchemasResponse
  );

  //assert
  expect(emitResult.errors).toHaveLength(0);
  expect(emitResult.prettifiedContent).toMatchSnapshot();
});

test("schema has immutable property", async () => {
  //arrange
  vi.setSystemTime(new Date(2023, 1, 1));

  const schemas: Array<Partial<Schema>> = [
    getTypedContextSchema(),
    {
      id: "SomeInterfaceName",
      properties: {
        immutableProperty: {
          type: "variable",
        },
      },
      immutable: ["immutableProperty"],
    },
  ];

  //act
  const emitResult = await emitToString(
    "4.13.8",
    "https://ftrack.example.com",
    schemas as QuerySchemasResponse
  );

  //assert
  expect(emitResult.errors).toHaveLength(0);
  expect(emitResult.prettifiedContent).toMatchSnapshot();
});

test("schema has integer type", async () => {
  //arrange
  vi.setSystemTime(new Date(2023, 1, 1));

  const schemas: Array<Partial<Schema>> = [
    getTypedContextSchema(),
    {
      id: "SomeInterfaceName",
      properties: {
        integerProperty: {
          type: "integer",
        },
      },
    },
  ];

  //act
  const emitResult = await emitToString(
    "4.13.8",
    "https://ftrack.example.com",
    schemas as QuerySchemasResponse
  );

  //assert
  expect(emitResult.errors).toHaveLength(0);
  expect(emitResult.prettifiedContent).toMatchSnapshot();
});

test("schema has variable type", async () => {
  //arrange
  vi.setSystemTime(new Date(2023, 1, 1));

  const schemas: Array<Partial<Schema>> = [
    getTypedContextSchema(),
    {
      id: "SomeInterfaceName",
      properties: {
        integerProperty: {
          type: "variable",
        },
      },
    },
  ];

  //act
  const emitResult = await emitToString(
    "4.13.8",
    "https://ftrack.example.com",
    schemas as QuerySchemasResponse
  );

  //assert
  expect(emitResult.errors).toHaveLength(0);
  expect(emitResult.prettifiedContent).toMatchSnapshot();
});

test("schema has array type", async () => {
  //arrange
  vi.setSystemTime(new Date(2023, 1, 1));

  const schemas: Array<Partial<Schema>> = [
    getTypedContextSchema(),
    {
      id: "SomeInterfaceName",
      properties: {
        arrayProperty: {
          type: "array",
          items: {
            $ref: "ArrayItem",
          },
        },
        mappedArrayProperty: {
          type: "mapped_array",
          items: {
            $ref: "ArrayItem",
          },
        },
      },
    },
    {
      id: "ArrayItem",
      type: "variable",
    },
  ];

  //act
  const emitResult = await emitToString(
    "4.13.8",
    "https://ftrack.example.com",
    schemas as QuerySchemasResponse
  );

  //assert
  expect(emitResult.errors).toHaveLength(0);
  expect(emitResult.prettifiedContent).toMatchSnapshot();
});

function getTypedContextSchema() {
  return {
    id: "TypedContext",
    properties: {},
  } as Partial<Schema> as Schema;
}
