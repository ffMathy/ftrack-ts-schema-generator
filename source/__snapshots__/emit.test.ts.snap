// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schema has array type 1`] = `
"// :copyright: Copyright (c) 2023 ftrack

// Generated on 2023-01-31T23:00:00.000Z using schema
// from an instance running version 4.13.8 using server on https://ftrack.example.com
// Not intended to modify manually

export interface TypedContext<
  K extends TypedContextSubtype = TypedContextSubtype
> {
  __entity_type__?: K;
  __permissions?: Record<string, any>;
}
export interface SomeInterfaceName {
  arrayProperty?: ArrayItem[];
  mappedArrayProperty?: ArrayItem[];
  __entity_type__?: \\"SomeInterfaceName\\";
  __permissions?: Record<string, any>;
}
export interface ArrayItem {
  __entity_type__?: \\"ArrayItem\\";
  __permissions?: Record<string, any>;
}
export interface BasicLink {
  id: string;
  type: string;
  name: string;
}
export interface EntityTypeMap {
  TypedContext: TypedContext;
  SomeInterfaceName: SomeInterfaceName;
  ArrayItem: ArrayItem;
}
export type EntityType = keyof EntityTypeMap;
export type EntityData<TEntityType extends EntityType = EntityType> =
  EntityTypeMap[TEntityType];
export interface TypedContextSubtypeMap {
  TypedContext: TypedContext;
}
export type TypedContextSubtype = keyof TypedContextSubtypeMap;

// Errors:
"
`;

exports[`schema has base schema 1`] = `
"// :copyright: Copyright (c) 2023 ftrack

// Generated on 2023-01-31T23:00:00.000Z using schema
// from an instance running version 4.13.8 using server on https://ftrack.example.com
// Not intended to modify manually

export interface TypedContext<
  K extends TypedContextSubtype = TypedContextSubtype
> {
  __entity_type__?: K;
  __permissions?: Record<string, any>;
}
export interface SomeInterfaceName
  extends Omit<SomeBaseSchema, \\"__entity_type__\\" | \\"__permissions\\"> {
  __entity_type__?: \\"SomeInterfaceName\\";
  __permissions?: Record<string, any>;
}
export interface SomeBaseSchema {
  __entity_type__?: \\"SomeBaseSchema\\";
  __permissions?: Record<string, any>;
}
export interface BasicLink {
  id: string;
  type: string;
  name: string;
}
export interface EntityTypeMap {
  TypedContext: TypedContext;
  SomeInterfaceName: SomeInterfaceName;
  SomeBaseSchema: SomeBaseSchema;
}
export type EntityType = keyof EntityTypeMap;
export type EntityData<TEntityType extends EntityType = EntityType> =
  EntityTypeMap[TEntityType];
export interface TypedContextSubtypeMap {
  TypedContext: TypedContext;
}
export type TypedContextSubtype = keyof TypedContextSubtypeMap;

// Errors:
"
`;

exports[`schema has immutable property 1`] = `
"// :copyright: Copyright (c) 2023 ftrack

// Generated on 2023-01-31T23:00:00.000Z using schema
// from an instance running version 4.13.8 using server on https://ftrack.example.com
// Not intended to modify manually

export interface TypedContext<
  K extends TypedContextSubtype = TypedContextSubtype
> {
  __entity_type__?: K;
  __permissions?: Record<string, any>;
}
export interface SomeInterfaceName {
  readonly immutableProperty?: string | number | boolean | string[];
  __entity_type__?: \\"SomeInterfaceName\\";
  __permissions?: Record<string, any>;
}
export interface BasicLink {
  id: string;
  type: string;
  name: string;
}
export interface EntityTypeMap {
  TypedContext: TypedContext;
  SomeInterfaceName: SomeInterfaceName;
}
export type EntityType = keyof EntityTypeMap;
export type EntityData<TEntityType extends EntityType = EntityType> =
  EntityTypeMap[TEntityType];
export interface TypedContextSubtypeMap {
  TypedContext: TypedContext;
}
export type TypedContextSubtype = keyof TypedContextSubtypeMap;

// Errors:
"
`;

exports[`schema has integer type 1`] = `
"// :copyright: Copyright (c) 2023 ftrack

// Generated on 2023-01-31T23:00:00.000Z using schema
// from an instance running version 4.13.8 using server on https://ftrack.example.com
// Not intended to modify manually

export interface TypedContext<
  K extends TypedContextSubtype = TypedContextSubtype
> {
  __entity_type__?: K;
  __permissions?: Record<string, any>;
}
export interface SomeInterfaceName {
  integerProperty?: number;
  __entity_type__?: \\"SomeInterfaceName\\";
  __permissions?: Record<string, any>;
}
export interface BasicLink {
  id: string;
  type: string;
  name: string;
}
export interface EntityTypeMap {
  TypedContext: TypedContext;
  SomeInterfaceName: SomeInterfaceName;
}
export type EntityType = keyof EntityTypeMap;
export type EntityData<TEntityType extends EntityType = EntityType> =
  EntityTypeMap[TEntityType];
export interface TypedContextSubtypeMap {
  TypedContext: TypedContext;
}
export type TypedContextSubtype = keyof TypedContextSubtypeMap;

// Errors:
"
`;

exports[`schema has variable type 1`] = `
"// :copyright: Copyright (c) 2023 ftrack

// Generated on 2023-01-31T23:00:00.000Z using schema
// from an instance running version 4.13.8 using server on https://ftrack.example.com
// Not intended to modify manually

export interface TypedContext<
  K extends TypedContextSubtype = TypedContextSubtype
> {
  __entity_type__?: K;
  __permissions?: Record<string, any>;
}
export interface SomeInterfaceName {
  integerProperty?: string | number | boolean | string[];
  __entity_type__?: \\"SomeInterfaceName\\";
  __permissions?: Record<string, any>;
}
export interface BasicLink {
  id: string;
  type: string;
  name: string;
}
export interface EntityTypeMap {
  TypedContext: TypedContext;
  SomeInterfaceName: SomeInterfaceName;
}
export type EntityType = keyof EntityTypeMap;
export type EntityData<TEntityType extends EntityType = EntityType> =
  EntityTypeMap[TEntityType];
export interface TypedContextSubtypeMap {
  TypedContext: TypedContext;
}
export type TypedContextSubtype = keyof TypedContextSubtypeMap;

// Errors:
"
`;

exports[`schema subtype of TypedContext 1`] = `
"// :copyright: Copyright (c) 2023 ftrack

// Generated on 2023-01-31T23:00:00.000Z using schema
// from an instance running version 4.13.8 using server on https://ftrack.example.com
// Not intended to modify manually

export interface TypedContext<
  K extends TypedContextSubtype = TypedContextSubtype
> {
  __entity_type__?: K;
  __permissions?: Record<string, any>;
}
export type SomeInterfaceName = TypedContext<\\"SomeInterfaceName\\">;
export interface BasicLink {
  id: string;
  type: string;
  name: string;
}
export interface EntityTypeMap {
  TypedContext: TypedContext;
  SomeInterfaceName: SomeInterfaceName;
}
export type EntityType = keyof EntityTypeMap;
export type EntityData<TEntityType extends EntityType = EntityType> =
  EntityTypeMap[TEntityType];
export interface TypedContextSubtypeMap {
  TypedContext: TypedContext;
  SomeInterfaceName: SomeInterfaceName;
}
export type TypedContextSubtype = keyof TypedContextSubtypeMap;

// Errors:
"
`;
